rfcs/0067-quarterly-releases.md

```diff
@@ -0,0 +1,66 @@
+    * The release manager begins gathering release notes
+
+* Two weeks prior to release
+    * Cherry picks become limited to critical changes only
+    * The release manager begins building binaries and testing them against TVM's test suite
+    * The release manager sends the release notes and plan to the PMC for approval (feedback thread will be open for one week)
+    * The release manager performs another license audit of all project dependencies to ensure they are compatible with Apache
+
+* One week prior to release
+    * The release manager sends the release and binaries to the PMC for an approval vote (thread open for one week). If rejected the building and test process must be started anew and repeated until the PMC approves.
+
+* Day of the release
+    * The release manager publishes the relevant binaries
+    * The release manager closes the release issue
+    * The release manager makes a GitHub release and updates an in-repo file `RELEASE.md` on both `main` and the release branch with the release notes
+
+Much of this can be automated via GitHub Actions on the apache/tvm repo. Eventually (though maybe not for the upcoming release) these will handle all the building, testing, and publishing of releases so the job of the release manager will become simpler over time.
+
+The release manager will use the release notes and discussions with developers to determine the next version number. Releases will continue the current versioning scheme of `major.minor.patch`, with a typical release involving a bump of the minor release version. Patch versions will be used for follow up releases onto a quarterly release, but not for the next quarter's release.
+
+This RFC will not go into effect until we resolve the API around `relay.build` since that may introduce significant churn in the community. See [the discussion](https://discuss.tvm.apache.org/t/pre-rfc-compilation-configuration-representation/11372) for details.
+
+Even with this RFC the ultimate decision to releases rests with the community. If no one wants to do a release, then it will be skipped or delayed until the community agrees and a release manager can be selected.
+
+# Prior Work
+
+* This [proposal](https://discuss.tvm.apache.org/t/pre-rfc-switch-to-time-based-releases/4245) suggests a similar thing for TVM and links to a great [write-up](https://cwiki.apache.org/confluence/display/KAFKA/Time+Based+Release+Plan) from the Kafka project describing their plan.
+
+# Drawbacks
+
+* This requires a high level of commitment from a single individual (the release manager). If someone is not able to fill or execute this role effectively the release will be stalled.
+
+# Rationale and alternatives
+
+* The main alternative is waiting until specific features have been developed in order to make a release. This has caveats as TVM has many sub-projects within it, so it's not clear which is significant enough to warrant a release.
```

Comment:
On this point of "_waiting until specific features have been developed in order to make a release_", I think as a very-active project, we always have quite a few new features being developed, so I don't think we should have as a general approach to *wait* for features to do releases.

In my opinion, we should have as a general policy that any major feature should have a toggle, to be disabled by default while it is in development stage, so that releases can be taken without partial/broken features in it.

Reply:
both good points though I would say using feature flags and how sub-teams do planning is outside the scope of this RFC. In general nothing should ever be broken on `main` and we can hopefully establish some processes to ensure API compatibility doesn't break in crazy ways between releases as part of the release testing

Reply:
Hmm, I am not sure I agree fully with the fact that this is out of scope for the RFC.  

This RFC is to set out the policy and set out a timeline for releases. 

One of the current drawbacks of the TVM project is that industrial contributors find it difficult to see how their customers would consume the open source project without the existing standard basic practices of regular releases and release trains. 

The consequence of that is that industrial contributors use these timelines to plan for their activities and would adjust other activities they have to do in order to make sure that features are landed in a sensible fashion . If TVM needs to remain relevant to industry, then these fundamental software engineering practices need to be put in place and remain.

Reply:
> both good points though I would say using feature flags and how sub-teams do planning is outside the scope of this RFC.

How teams will organise to implement such feature toggles is their decision - agreed.

To be more specific, what I think worth to be explicit in this RFC is that such mechanisms should be mandatory, as a way to keep viable releases going. Often - for technical reasons - features are still making their way into the repository, so it would be good to create a culture in which even with partial features in the repo, we can still release if we need/want to.